File: tests/accesses/scale/accesses_scale.ml
open Optitrust
open Prelude

let _ = Flags.check_validity := true
let _ = Flags.recompute_resources_between_steps := true

let _ = Run.script_cpp (fun _ ->

  !! Accesses.scale_var ~factor:(trm_float 5.0) [cTopFunDef "test_var"; cVarDef "x"];
  !! Accesses.scale_immut ~factor:(trm_float 3.0) [cTopFunDef "test_var"; cVarDef "z"];

  let t = trm_find_var "t" [cTopFunDef "test_array"] in
  let address_pattern = trm_array_access t (trm_int 0) in
  !! Accesses.scale ~factor:(trm_float 5.0) ~address_pattern [tSpanSeq [cFunBody "test_array"]];
(*  TODO : Mettre exemple minimal scale qui marche pas  *)
  (* TODO:
  !! Accesses.scale ~factor:(trm_float 5.0) [nbMulti; cTopFunDef "main"; sExpr "t[i]"]; *)
)

File: tests/accesses/scale/accesses_scale.cpp
#include <optitrust.h>

void test_var() {
  __pure();

  double x = 0.0;
  const double z = 1.0;
  x = 1.0 + z;
  double y = x + 1.0;
  x = x + 2.0;
}

void test_array(double* t) {
  __reads("&t[2] ~> Cell");
  __modifies("&t[0] ~> Cell, &t[1] ~> Cell");
  t[0] = t[0] + 1.0 + t[2];
}
/*
int main() {
  __pure();

  double t[3] = { 1., 2., 3. };
  double v = 2.0;
  int i = 0;
  t[i] = t[i] + v;
  return 0;
}
*/
